`brew upgrade` sometimes doesn't respect pinned dependency

<details>
<summary><h3><code>brew doctor</code> output</h3></summary>

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected header files:
  /usr/local/include/ezs.h
  /usr/local/include/ezsthread.h
  /usr/local/include/node/*
  /usr/local/include/python3.8/*

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libezs.la

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/python-3.8-embed.pc
  /usr/local/lib/pkgconfig/python-3.8.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libezs.a

Warning: Your /home directory is a symlink.
This is known to cause issues with formula linking, particularly when installing
multiple formulae that create symlinks in shared directories.

While this may be a standard directory structure in some distributions
(e.g. Fedora Silverblue) there are known issues as-is.

If you encounter linking issues, you may need to manually create conflicting
directories or use `brew link --overwrite` as a workaround.

We'd welcome a PR to fix this functionality.
See https://github.com/Homebrew/brew/issues/18036 for more context.

This is a Tier 2 configuration:
  https://docs.brew.sh/Support-Tiers#tier-2
You can report issues with Tier 2 configurations to Homebrew/* repositories!
Read the above document before opening any issues or PRs.


Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  binutils

Warning: Your system glibc 2.17 is too old.
We will need to automatically install a newer version.

We recommend updating to a newer version via your distribution's
package manager, upgrading your distribution to the latest version,
or changing distributions.

This is a Tier 2 configuration:
  https://docs.brew.sh/Support-Tiers#tier-2
You can report issues with Tier 2 configurations to Homebrew/* repositories!
Read the above document before opening any issues or PRs.


Warning: Your Homebrew's prefix is not /home/linuxbrew/.linuxbrew.

Most of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.

This is a Tier 3 configuration:
  https://docs.brew.sh/Support-Tiers#tier-3
You can report issues with Tier 3 configurations to Homebrew/* repositories!
Read the above document before opening any issues or PRs.


Warning: /usr/bin occurs before /data/shared/lusers/jacktose/homebrew/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/data/shared/lusers/jacktose/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/data/shared/lusers/jacktose/homebrew/bin:$PATH"' >> /usr/lusers/jacktose/.bash_profile

The following tools exist at both paths:
  attr
  autopoint
  awk
  bash
  bashbug
  bison
  bunzip2
  bzcat
  bzcmp
  bzdiff
  bzgrep
  bzip2
  bzip2recover
  bzless
  bzmore
  cal
  captoinfo
  chacl
  chmem
  chrt
  clear
  col
  colcrt
  colrm
  column
  compile_et
  corelist
  cpan
  curl
  curl-config
  dmesg
  eject
  envsubst
  fallocate
  findmnt
  flex
  flex++
  flock
  funzip
  gawk
  getfacl
  getfattr
  getopt
  gettext
  gettext.sh
  gettextize
  gfortran
  gss-client
  h2ph
  h2xs
  hexdump
  i386
  infocmp
  infotocap
  instmodsh
  ionice
  ipcmk
  ipcrm
  ipcs
  isosize
  json_pp
  k5srvutil
  kadmin
  kdestroy
  keyctl
  kinit
  klist
  kpasswd
  krb5-config
  ksu
  kswitch
  ktutil
  kvno
  last
  lastb
  ldapadd
  ldapcompare
  ldapdelete
  ldapexop
  ldapmodify
  ldapmodrdn
  ldappasswd
  ldapsearch
  ldapurl
  ldapwhoami
  less
  lesskey
  lex
  libnetcfg
  linux32
  linux64
  logger
  look
  lsblk
  lscpu
  lsipc
  lslocks
  lslogins
  lsmem
  lsns
  lz4
  lz4c
  lz4cat
  m4
  make
  mcookie
  mesg
  more
  mount
  mountpoint
  msgattrib
  msgcat
  msgcmp
  msgcomm
  msgconv
  msgen
  msgexec
  msgfilter
  msgfmt
  msggrep
  msginit
  msgmerge
  msgunfmt
  msguniq
  namei
  ngettext
  nsenter
  openssl
  perl
  perlbug
  perldoc
  perlivp
  perlthanks
  piconv
  pkg-config
  pl2pm
  pod2html
  pod2man
  pod2text
  pod2usage
  podchecker
  prlimit
  prove
  ptar
  ptardiff
  ptargrep
  python3
  recode-sr-latin
  rename
  renice
  reset
  rev
  script
  scriptreplay
  setarch
  setfacl
  setfattr
  setsid
  setterm
  shasum
  sim_client
  splain
  sqlite3
  tabs
  taskset
  tic
  toe
  tput
  tset
  tzselect
  ul
  umount
  unlz4
  unshare
  unxz
  unzip
  unzipsfx
  utmpdump
  uuclient
  uuidgen
  wall
  wdctl
  whereis
  whois
  x86_64
  xgettext
  xml2-config
  xmlcatalog
  xmllint
  xmlwf
  xsubpp
  xz
  xzcat
  xzcmp
  xzdec
  xzdiff
  xzegrep
  xzfgrep
  xzgrep
  xzless
  xzmore
  yacc
  zipdetails
  zipgrep
  zipinfo

Warning: Homebrew's "bin" was not found in your PATH.
Consider setting your PATH for example like so:
  echo 'export PATH="/data/shared/lusers/jacktose/homebrew/bin:$PATH"' >> /usr/lusers/jacktose/.bash_profile

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /data/shared/lusers/jacktose/homebrew/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/data/shared/lusers/jacktose/homebrew/sbin:$PATH"' >> /usr/lusers/jacktose/.bash_profile
```
</details>

### Verification

- [x] I ran `brew update` twice and am still able to reproduce my issue.
- [ ] My "`brew doctor` output" above says `Your system is ready to brew` or a definitely unrelated `Tier` message.
    - I believe the messages are all unrelated. I might be able to replicate on a cleaner machine if necessary, but I think this is clearly `brew` logic & UI, not a machine issue.
- [x] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

<details>
<summary><h3><code>brew config</code> output</h3></summary>

```shell
HOMEBREW_VERSION: 6.0.6-2-g7f14ed4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7f14ed41545a48dd288e2e0b0e940bf6325aa47a
Last commit: 10 hours ago
Branch: main
Core tap JSON: 01 Jul 17:26 UTC
HOMEBREW_PREFIX: /data/shared/lusers/jacktose/homebrew
HOMEBREW_REPOSITORY: /data/shared/lusers/jacktose/homebrew
HOMEBREW_CELLAR: /data/shared/lusers/jacktose/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_CLEANUP_MAX_AGE_DAYS: 60
HOMEBREW_CLEANUP_PERIODIC_FULL_DAYS: 7
HOMEBREW_DISPLAY_INSTALL_TIMES: set
HOMEBREW_DOWNLOAD_CONCURRENCY: 8
HOMEBREW_EDITOR: vim
HOMEBREW_FORBID_PACKAGES_FROM_PATHS: set
HOMEBREW_FORCE_BREWED_CURL: set
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_GIT_PATH: /usr/local/bin/git
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_SANDBOX_LINUX: set
HOMEBREW_REQUIRE_TAP_TRUST: set
Homebrew Ruby: 4.0.5 => /data/shared/lusers/jacktose/homebrew/Library/Homebrew/vendor/portable-ruby/4.0.5_1/bin/ruby
CPU: quad-core 64-bit broadwell
Clang: N/A
Git: 2.31.1 => /usr/local/bin/git
Curl: 8.15.0 => /data/shared/lusers/jacktose/homebrew/opt/curl/bin/curl
Kernel: Linux 3.10.0-1160.90.1.el7.x86_64 x86_64 GNU/Linux
OS: CentOS Linux release 7.9.2009 (Core) (Core)
Host glibc: 2.17
Host libstdc++: N/A
/usr/bin/gcc: 4.8.5
/usr/bin/ruby: N/A
glibc: 2.39
gcc@13: 15.3.0
gcc: 15.3.0
xorg: N/A
```
</details>

### What were you trying to do (and why)?

Have gcc (and some others) pinned because the latest version is failing to build.
Upgrade whatever I can that doesn't depend on gcc.

### What happened (include all command output)?

```shell
$ brew list --pinned --versions
gawk 5.3.1
gcc 15.3.0
openldap 2.6.10
python@3.13 3.13.7

brew info libffi
==> libffi ↑: 3.5.2 → stable 3.6.0 (bottled), HEAD
Portable Foreign Function Interface library
https://sourceware.org/libffi/
Installed (as dependency)
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/lib/libffi.rb
License: MIT
==> Installed Kegs and Versions
libffi ↑ 3.5.2 → 3.6.0 (19 files, 805.8KB) [Linked]
==> Dependencies
Required (2): gcc ↑, glibc ✔
Recursive Runtime (2): all installed ✔
Dependents: 2
...

$ brew info gcc
==> gcc ↑: 15.3.0 → stable 16.1.0, HEAD
GNU compiler collection
https://gcc.gnu.org/
Aliases: gcc@16, gfortran
Installed (on request)
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/g/gcc.rb
License: GPL-3.0-or-later WITH GCC-exception-3.1
Pinned: 15.3.0 on 2026-06-29 at 15:14:01
==> Installed Kegs and Versions
gcc ↑ 15.3.0 → 16.1.0 (2,173 files, 458.6MB) [Linked]
==> Dependencies
Required (8): gmp ✔, isl ✔, libmpc ✔, mpfr ✔, zstd ✔, binutils ✔, zlib-ng-compat ✔, glibc ✔
Recursive Runtime (11): all installed ✔
Dependents: 44
...

$ brew upgrade
✔︎ JSON API packages.x86_64_linux.jws.json
Error: You must `brew unpin gcc` as installing libnghttp3 requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing libngtcp2 requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing sqlite requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc openldap` as installing curl requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing expat requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing uv requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing oh-my-posh requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing llhttp requires the latest version of pinned dependencies.
Warning: Not upgrading 4 pinned packages:
Warning: Not upgrading 4 pinned packages:
Error: You must `brew unpin gcc` as installing libnghttp3 requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing libngtcp2 requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing sqlite requires the latest version of pinned dependencies.
gcc 16.1.0, openldap 2.6.13, gawk 5.4.0, python@3.13 3.13.14
==> Would install 1 dependency for libffi:
gcc 15.3.0 -> 16.1.0
==> Would upgrade 1 outdated package
libffi 3.5.2 -> 3.6.0 (284.7KB)
==> 4 Pinned formulae
gcc 16.1.0
openldap 2.6.13
gawk 5.4.0
python@3.13 3.13.14
==> Upgrading 1 outdated package:
libffi 3.5.2 -> 3.6.0 (284.7KB)
gcc 16.1.0, openldap 2.6.13, gawk 5.4.0, python@3.13 3.13.14
Error: You must `brew unpin gcc openldap` as installing curl requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing expat requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing uv requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing oh-my-posh requires the latest version of pinned dependencies.
Error: You must `brew unpin gcc` as installing llhttp requires the latest version of pinned dependencies.
Warning: Building gcc from source as the bottle needs:
- `HOMEBREW_CELLAR=/home/linuxbrew/.linuxbrew/Cellar` (yours is /data/shared/lusers/jacktose/homebrew/Cellar)
- `HOMEBREW_PREFIX=/home/linuxbrew/.linuxbrew` (yours is /data/shared/lusers/jacktose/homebrew)
✔︎ API Source gcc.rb
✔︎ Bottle libffi (3.6.0)
✔︎ Formula gcc (16.1.0)
==> Fetching downloads for: libffi
==> Upgrading libffi
  3.5.2 -> 3.6.0 
==> Installing libffi dependency: gcc
Warning: Sandbox unavailable: building without sandboxing!
==> ../configure --prefix=/data/shared/lusers/jacktose/homebrew/opt/gcc --libdir=/data/shared/lusers/jacktose/homebrew/opt/gcc/lib/gcc/current --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran,m2 --program-suffix=-16 --with-gmp=/data/shared/lusers/jacktose/homebrew/opt/gmp --with-mpfr=/data/shared/lusers/jacktose/homebrew/opt/mpfr --with-mpc=/data/shared/lusers/jacktose/homebrew/opt/libmpc --with-isl=/data/shared/lusers/jacktose/homebrew/opt/isl --with-zstd=/data/shared/lusers/jacktose/homebrew/opt/zstd --with-pkgversion=Homebrew GCC 16.1.0 --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --with-system-zlib --disable-multilib --enable-default-pie --with-as=/data/shared/lusers/jacktose/homebrew/opt/binutils/bin/as
Last 15 lines from /usr/lusers/jacktose/.cache/Homebrew/Logs/gcc/01.configure.log:
checking whether ln works... yes
checking whether ln -s works... yes
checking for a sed that does not truncate output... /usr/bin/sed
checking for gawk... gawk
checking for libatomic support... yes
checking for libitm support... yes
checking for libsanitizer support... yes
checking for libvtv support... yes
checking for libphobos support... yes
checking for libgcobol support... yes
checking for gcc... gcc-15
checking whether the C compiler works... no
configure: error: in `/var/tmp/gcc-20260701-1021-7oq3x9/gcc-16.1.0/build':
configure: error: C compiler cannot create executables
See `config.log' for more details

READ THIS: https://docs.brew.sh/Troubleshooting

These open issues may also help:
gcc 16.1.0 bottle ships empty libstdc++ std module interface files on Apple Silicon https://github.com/Homebrew/homebrew-core/issues/289142

==> 4 Pinned formulae
gcc 16.1.0
openldap 2.6.13
gawk 5.4.0
python@3.13 3.13.14
```

### What did you expect to happen?

Expected `brew upgrade` to skip libffi because of its pinned dependency, as it did for 8 other formulae.
Instead, the dry-run/ask output lists “would install 1 dependency: gcc” and “would upgrade 1 outdated: libffi” and then it proceeds to try to install gcc.

I had a similar issue with pinned dependency logic in #22676 (fixed in #22683), but that was only the dry-run/ask output; this is for that output plus the actual install behavior; but it only affects some packages.

Maybe there's something weird with [libffi]'s dependency tree? It doesn't list any deps in the formula, so I don't know where the gcc dep comes from. It's also not listed in `brew deps`, but it is in `brew deps --tree`:
<details>
<summary><code>brew deps</code></summary>

```shell
$ brew deps libffi
glibc
linux-headers@5.15

brew deps libffi --tree
Warning: `brew deps` is not the actual runtime dependencies because --tree was passed!
This means dependencies may differ from a formula's declared dependencies.
Hide these hints with `HOMEBREW_NO_ENV_HINTS=1` (see `man brew`).
libffi
├── gcc
│   ├── gmp
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── isl
│   │   ├── gmp
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── libmpc
│   │   ├── gmp
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   ├── mpfr
│   │   │   ├── gmp
│   │   │   │   └── glibc
│   │   │   │       └── linux-headers@6.8
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── mpfr
│   │   ├── gmp
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── zstd
│   │   ├── lz4
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   ├── xz
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   ├── zlib-ng-compat
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── binutils
│   │   ├── zstd
│   │   │   ├── lz4
│   │   │   │   └── glibc
│   │   │   │       └── linux-headers@6.8
│   │   │   ├── xz
│   │   │   │   └── glibc
│   │   │   │       └── linux-headers@6.8
│   │   │   ├── zlib-ng-compat
│   │   │   │   └── glibc
│   │   │   │       └── linux-headers@6.8
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   ├── zlib-ng-compat
│   │   │   └── glibc
│   │   │       └── linux-headers@6.8
│   │   └── glibc
│   │       └── linux-headers@6.8
│   ├── zlib-ng-compat
│   │   └── glibc
│   │       └── linux-headers@6.8
│   └── glibc
│       └── linux-headers@6.8
└── glibc
    └── linux-headers@6.8
```
</details>

Or maybe there's something wrong with the gcc version logic? I notice that the `brew upgrade` output says `gcc 16.1.0`, but the installed & pinned version is `gcc 15.3.0`.

[libffi]: https://github.com/Homebrew/homebrew-core/blob/c3e11bcd9a6ae4a5cd7caae32324857c97e49257/Formula/lib/libffi.rb

### Step-by-step reproduction instructions (running `brew` commands)

```shell
brew install gcc libffi [at gcc v. 15.3.0 & libffi v. 3.5.2]
brew pin gcc
[wait for libffi update]
brew upgrade
```

--- a/Library/Homebrew/formula_installer.rb
+++ b/Library/Homebrew/formula_installer.rb
@@ -474,8 +474,7 @@ class FormulaInstaller
 
     recursive_deps = if pour_bottle?
       # Include implicit dependencies (except duplicates) in formulae to check
-      active_spec = T.must(formula.stable? ? formula.stable : formula.head)
-      (formula.runtime_dependencies + active_spec.deps.required).uniq(&:name)
+      (formula.runtime_dependencies + formula.deps.select(&:implicit?)).uniq(&:name)
     else
       formula.recursive_dependencies
     end

--- a/Library/Homebrew/formula_installer.rb
+++ b/Library/Homebrew/formula_installer.rb
@@ -473,7 +473,9 @@ class FormulaInstaller
     end
 
     recursive_deps = if pour_bottle?
-      (formula.runtime_dependencies + (formula.stable&.deps&.required || [])).uniq(&:name)
+      # Include implicit dependencies (except duplicates) in formulae to check
+      active_spec = T.must(formula.stable? ? formula.stable : formula.head)
+      (formula.runtime_dependencies + active_spec.deps.required).uniq(&:name)
     else
       formula.recursive_dependencies
     end

--- a/Library/Homebrew/formula_installer.rb
+++ b/Library/Homebrew/formula_installer.rb
@@ -473,7 +473,7 @@ class FormulaInstaller
     end
 
     recursive_deps = if pour_bottle?
-      formula.runtime_dependencies
+      (formula.runtime_dependencies + (formula.stable&.deps&.required || [])).uniq(&:name)
     else
       formula.recursive_dependencies
     end

